### PR TITLE
[Merged by Bors] - standardize account balance to uint64

### LINKED ISSUE #2192: Standardize account balance type
## Description

In some places we use uint64 for balance, in other places we use big.Int. Let's standardize all of these into a single type, uint64.

This issue appears in commit hash: e353ccc393eda807411ef001015b2d3e8dabf4fb

See also #2069

### PR DESCRIPTION
**Pull request by sidsankhe**
## Motivation

Closes #2192 

## Changes

Changed the genesis account structs and the account_state to use uint64 instead of big.Int. Changed the associated method signatures and tests as well. 
## Test Plan

Since there were no functionality changes, no new tests were added. All the existing tests were updated to comply with the new types.
## TODO

- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update documentation as needed

### CODE REVIEW

**Comment by lrettig** [comment]
Great start :) Please give this PR a more descriptive title when you get a chance!

File: api/config/genesis.go [hunk 1]
```diff
@@ -3,16 +3,15 @@ package config
...
 	"os"
+
+	"github.com/spacemeshos/go-spacemesh/log"
 )
 
 // GenesisAccount is the json representation of an account
 type GenesisAccount struct {
-	Balance *big.Int `json:"balance" gencodec:"required"`
-	Nonce   uint64   `json:"nonce"`
+	Balance uint64 `json:"balance" gencodec:"required"`
```
Comment (lrettig) [inline api/config/genesis.go]: Is this still `required` now that it's no longer a pointer?

File: api/config/genesis.go [hunk 2]
```diff
@@ -75,9 +74,9 @@ func DefaultGenesisConfig() *GenesisConfig {
 
 	// we default to 10^5 SMH per account which is 10^17 smidge
 	g.InitialAccounts = map[string]GenesisAccount{
-		"0x1":       {Balance: big.NewInt(int64(math.Pow10(17))), Nonce: 0},
-		Account1Pub: {Balance: big.NewInt(int64(math.Pow10(17))), Nonce: 0},
-		Account2Pub: {Balance: big.NewInt(int64(math.Pow10(17))), Nonce: 0},
+		"0x1":       {Balance: 100000000000000000, Nonce: 0},
```
Comment (lrettig) [inline api/config/genesis.go:78]: It's now a little less clear to the naked eye what this value is. Could you add a comment? Something like, "Genesis balance is 10^17"

File: state/processor.go [hunk 3]
```diff
@@ -187,17 +188,18 @@ func (tp *TransactionProcessor) GetLayerStateRoot(layer types.LayerID) (types.Ha
 
 // ApplyRewards applies reward reward to miners vector miners in for layer
 func (tp *TransactionProcessor) ApplyRewards(layer types.LayerID, miners []types.Address, reward *big.Int) {
+	rewardConverted := reward.Uint64()
```
Comment (lrettig) [inline state/processor.go:192]: Please add a "TODO" comment here with a link to the issue where this will be fixed (#2069)

**Review by lrettig — CHANGES_REQUESTED** [review]
Looking good! See comments I added. One more thing, is `math/big` still being used anywhere outside the rewards code?

File: api/config/genesis.go (hunk 1 above)
Comment (sidsankhe) [reply api/config/genesis.go]: As mentioned, we should ask Noam to double check? Leaving this in for now.

**Comment by sidsankhe** [comment]
Output from the grepping math/big on the codebase:
`./mesh/mesh_test.go:    "math/big"
./mesh/mesh.go: "math/big"
./mesh/meshdb_test.go:  "math/big"
./mesh/reward_test.go:  "math/big"
./mesh/meshdb.go:       "math/big"
./mesh/reward.go:       "math/big"
./cmd/hare/oracle_client.go:    "math/big"
./activation/activationdb_test.go:      "math/big"
./state/account_state.go:       "math/big"
./state/processor_test.go:      "math/big"
./state/processor.go:   "math/big"
./rlp/decode_test.go:   "math/big"
./rlp/encode.go:        "math/big"
./rlp/decode.go:        "math/big"
./rlp/encode_test.go:   "math/big"
./common/types/address.go:      "math/big"
./common/types/hashes.go:       "math/big"
./common/types/big.go:import "math/big"
./common/util/hexutil.go:       "math/big"
./common/util/hexutil_test.go:  "math/big"
./common/util/json_test.go:     "math/big"
./common/util/json.go:  "math/big"
./trie/trie_test.go:    "math/big"
./api/grpcserver/grpcserver_test.go:    "math/big"
./sync/mocks.go:        "math/big"
./sync/syncer_test.go:  "math/big"`

Let me know if I'm missing anything.

**Comment by lrettig** [comment]
> Output from the grepping math/big on the codebase:

Thanks for posting this! IIRC when we spoke today you said you had reviewed these remaining usages of bigint and you didn't think we needed to change/remove them. Can you remind me why that's the case? At first glance it looks to me like a few or most of these should go - not that we have to address them all in this issue/PR, but we should make sure there are open issues for remaining usages that need to be updated. Maybe we could turn this list of files into a checklist and group them by remaining data structures or use cases, something along those lines.

File: state/processor.go [hunk 4]
```diff
@@ -187,6 +187,7 @@ func (tp *TransactionProcessor) GetLayerStateRoot(layer types.LayerID) (types.Ha
 }
 
 // ApplyRewards applies reward reward to miners vector miners in for layer
+// TODO: convert the rewards to uint64 (#2069)
```
Comment (lrettig) [inline state/processor.go]:
```suggestion
// TODO: convert rewards to uint64 (see https://github.com/spacemeshos/go-spacemesh/issues/2069)
```

**Review by lrettig — CHANGES_REQUESTED** [review]

File: api/config/genesis.go (hunk 1 above)
Comment (lrettig) [reply api/config/genesis.go]: This appears to be Anton's code from a while ago (#491). It doesn't appear to be used anywhere else in the codebase or mentioned anywhere in the docs so I think it's safe to remove the entire `gencodec` statement. @antonlerner let us know if you disagree/if we've missed something here!

Comment (lrettig) [reply api/config/genesis.go]: For reference: https://godoc.org/github.com/fjl/gencodec

**Comment by sidsankhe** [comment]
bors try

**Comment by sidsankhe** [comment]
bors try

**Comment by sidsankhe** [comment]
bors try

**Comment by lrettig** [comment]
@sidsankhe update the base branch and then I think this is good to go, well done :)

**Comment by lrettig** [comment]
> > Output from the grepping math/big on the codebase:
> 
> Thanks for posting this! IIRC when we spoke today you said you had reviewed these remaining usages of bigint and you didn't think we needed to change/remove them. Can you remind me why that's the case? At first glance it looks to me like a few or most of these should go - not that we have to address them all in this issue/PR, but we should make sure there are open issues for remaining usages that need to be updated. Maybe we could turn this list of files into a checklist and group them by remaining data structures or use cases, something along those lines.

We discussed this offline and agreed that, as all balance-related types have been updated (with the exception of the rewards stuff that will be done in #2069), other usages of bigint are out of scope for this issue/PR.

**Review by lrettig — APPROVED** [review]
LGTM! 😄

**Comment by sidsankhe** [comment]
bors merge

**Comment by sidsankhe** [comment]
bors merge

**Comment by sidsankhe** [comment]
bors merge

**Comment by sidsankhe** [comment]
bors merge

**Comment by sidsankhe** [comment]
bors merge

**Comment by sidsankhe** [comment]
bors merge

**Pull request closed without merging**